MonoGame.Framework/Graphics/Vertices/IndexBuffer.cs

```diff
@@ -115,10 +115,26 @@ internal protected override void GraphicsDeviceResetting()
         protected void SetDataInternal<T>(int offsetInBytes, T[] data, int startIndex, int elementCount, SetDataOptions options) where T : struct
         {
             if (data == null)
-                throw new ArgumentNullException("data is null");
+                throw new ArgumentNullException("data");
             if (data.Length < (startIndex + elementCount))
                 throw new InvalidOperationException("The array specified in the data parameter is not the correct size for the amount of data requested.");
 
+            var type = typeof (T);
+            switch (IndexElementSize)
+            {
+                case IndexElementSize.SixteenBits:
+                    if (type != typeof(short) || type != typeof(ushort))
+                    {
+                        throw new ArrayTypeMismatchException("Invalid type using when short or ushort need to be used.");
+                    }
+                    break;
+                case IndexElementSize.ThirtyTwoBits:
+                    if (type != typeof(int) || type != typeof(uint))
+                    {
+                        throw new ArrayTypeMismatchException("Invalid type using when int or uint need to be used.");
+                    }
+                    break;
+            }

```

Comment:
> I doubt if someone put exotic things like Color or float to IndexBuffer.

It isn't about doubt or what is logical...  it is what XNA supports.

I just tested this...

```
            var savedData = new short[] { 1, 2, 3, 4 };
            var asBytes = new byte[savedData.Length * sizeof(short)];
            Buffer.BlockCopy(savedData, 0, asBytes, 0, asBytes.Length);
            var indexBuffer = new IndexBuffer(GraphicsDevice, IndexElementSize.SixteenBits, savedData.Length, BufferUsage.None);
            indexBuffer.SetData(asBytes);
            var readData = new short[4];
            indexBuffer.GetData(readData, 0, 4);
            Assert.AreEqual(savedData, readData);
```

That works on XNA.  It would fail under your change.